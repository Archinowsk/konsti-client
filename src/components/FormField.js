// @flow
import React, { Fragment } from 'react';
import { useTranslation } from 'react-i18next';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { FieldProps } from 'redux-form';
import type { StatelessFunctionalComponent, Element } from 'react';

export const FormField: StatelessFunctionalComponent<FieldProps> = (
  props: FieldProps
): Element<typeof Fragment> => {
  // $FlowFixMe: Cannot get field.type because property type is missing in FieldProps [1].
  const { type } = props;
  const { name, onChange, onBlur, onDragStart, onDrop, onFocus } = props.input;
  const { touched, error } = props.meta;
  const { t } = useTranslation();

  const [fieldType, setFieldType] = React.useState('');
  (fieldType: string);

  React.useEffect(() => {
    setFieldType(type);
  }, [type]);

  const classNames = ['form-input'];

  if (type === 'checkbox') {
    classNames.push('checkbox');
  }

  const togglePasswordVisibility = () => {
    if (fieldType === 'password') setFieldType('text');
    else if (fieldType === 'text') setFieldType('password');
  };

  return (
    <Fragment>
      <div className='form-row'>
        <div className='form-field'>
          <input
            className={classNames.join(' ')}
            id={name}
            label={t(name)}
            name={name}
            onBlur={onBlur}
            onChange={onChange}
            onDragStart={onDragStart}
            onDrop={onDrop}
            onFocus={onFocus}
            placeholder={t(name)}
            type={fieldType}
          />
          {type === 'checkbox' && <label htmlFor={name}>{t(name)}</label>}
        </div>

        {type === 'password' && (
          <span className='form-field-icon'>
            <FontAwesomeIcon
              icon={fieldType === 'password' ? 'eye' : 'eye-slash'}
              onClick={togglePasswordVisibility}
            />
          </span>
        )}
      </div>

      {touched && error && (
        <div className='form-field-error'>
          <span className='form-field-error-message'>{t(error)}</span>
        </div>
      )}
    </Fragment>
  );
};
