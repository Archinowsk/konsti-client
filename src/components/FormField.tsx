import React, { FC, ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { FieldProps } from 'redux-form';

export const FormField: FC<FieldProps> = (props: FieldProps): ReactElement => {
  const { type } = props;
  const { name, onChange, onBlur, onDragStart, onDrop, onFocus } = props.input;
  const { touched, error } = props.meta;
  const { t } = useTranslation();

  const [fieldType, setFieldType] = React.useState<string>('');

  React.useEffect(() => {
    setFieldType(type);
  }, [type]);

  const classNames = ['form-input'];

  if (type === 'checkbox') {
    classNames.push('checkbox');
  }

  const togglePasswordVisibility = () => {
    if (fieldType === 'password') setFieldType('text');
    else if (fieldType === 'text') setFieldType('password');
  };

  return (
    <>
      <div className='form-row'>
        <div className='form-field'>
          <input
            className={classNames.join(' ')}
            id={name}
            name={name}
            onBlur={onBlur}
            onChange={onChange}
            onDragStart={onDragStart}
            onDrop={onDrop}
            onFocus={onFocus}
            placeholder={t(name)}
            type={fieldType}
          />
          {type === 'checkbox' && <label htmlFor={name}>{t(name)}</label>}
        </div>

        {type === 'password' && (
          <span className='form-field-icon'>
            <FontAwesomeIcon
              icon={fieldType === 'password' ? 'eye' : 'eye-slash'}
              onClick={togglePasswordVisibility}
            />
          </span>
        )}
      </div>

      {touched && error && (
        <div className='form-field-error'>
          <span className='form-field-error-message'>{t(error)}</span>
        </div>
      )}
    </>
  );
};
