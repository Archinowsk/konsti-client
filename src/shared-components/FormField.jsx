import React from 'react';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

const FormField = props => {
  const {
    input,
    label = '',
    type,
    required = false,
    meta: { touched, error },
    t,
    tooltip = true,
  } = props;

  return (
    <div className="form-row">
      <div className="form-field">
        {!required &&
          <label htmlFor={label}>
            {label}
          </label>}
        {required === true &&
          <label htmlFor={label}>
            {label} *
          </label>}

        <div className="form-input-wrapper" id={`${input.name}-wrapper`}>
          <input
            className="form-input"
            {...input}
            placeholder={label}
            type={type}
            id={input.name}
          />
        </div>

        {/*
        {tooltip === true &&
          <div className="help-tip">
            <p>
              {t(`tooltip.${input.name}`)}
            </p>
          </div>}
        */}
      </div>

      {touched &&
        error &&
        <div className="form-field-error">
          {error}
        </div>}
    </div>
  );
};

FormField.propTypes = {
  input: PropTypes.object.isRequired,
  label: PropTypes.string, // eslint-disable-line react/require-default-props
  type: PropTypes.string.isRequired,
  meta: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  required: PropTypes.bool, // eslint-disable-line react/require-default-props
  tooltip: PropTypes.bool, // eslint-disable-line react/require-default-props
};

export default translate()(FormField);
