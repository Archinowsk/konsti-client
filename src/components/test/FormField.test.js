/* @flow */
import React from 'react'
import { shallow } from 'enzyme'
import FormField from '../FormField'
import type { FieldProps } from 'redux-form'

describe('FormField', () => {
  it('should render correctly', () => {
    const props: FieldProps = {
      meta: {
        active: false,
        asyncValidating: false,
        autofilled: false,
        dirty: false,
        dispatch: () => {},
        error: 'validation.required',
        form: 'registration',
        initial: undefined,
        invalid: true,
        pristine: true,
        submitFailed: true,
        submitting: false,
        touched: true,
        valid: false,
        visited: false,
        warning: undefined,
      },
      input: {
        name: 'password',
        value: '',
        onBlur: () => {},
        onChange: () => {},
        onDragStart: () => {},
        onDrop: () => {},
        onFocus: () => {},
      },
      custom: {},
    }
    const component = shallow(<FormField {...props} />)
    expect(component).toMatchSnapshot()
  })
})
