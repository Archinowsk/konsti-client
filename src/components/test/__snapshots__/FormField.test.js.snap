// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should render correctly 1`] = `
<div
  className="form-row"
>
  <div
    className="form-field"
  >
    <input
      className="form-input"
      id="password"
      label="password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="password"
      type=""
      value=""
    />
  </div>
  <div
    className="form-field-error"
  >
    validation.required
  </div>
</div>
`;
